idx/memory/tag_query.go

```diff
@@ -171,247 +248,751 @@ func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
+			continue
+		}
+
+		// generate a set of all tags of the current metric that satisfy the
+		// tag filter condition
+		metricTags := make(map[string]struct{}, 0)
+		for _, tag := range def.Tags {
+			equal := strings.Index(tag, "=")
+			if equal < 0 {
+				corruptIndex.Inc()
+				log.Error(3, "memory-idx: tag is in index, but does not contain '=' sign: %s", tag)
+				continue
+			}
+
+			key := tag[:equal]
+			// this tag has already been pushed into tagCh, so we can stop evaluating
+			if _, ok := resultsCache[key]; ok {
+				continue
+			}
+
+			if q.filterTagBy == PREFIX_TAG {
+				if !strings.HasPrefix(key, q.tagPrefix) {
+					continue
+				}
+			} else if q.filterTagBy == MATCH_TAG {
+				if _, ok := q.tagMatch.missCache.Load(key); ok || !q.tagMatch.value.MatchString(tag) {
+					if !ok {
+						q.tagMatch.missCache.Store(key, struct{}{})
+					}
+					continue
+				}
+			}
+			metricTags[key] = struct{}{}
+		}
+
+		if matchName {
+			if _, ok := resultsCache["name"]; !ok {
+				metricTags["name"] = struct{}{}
+			}
+		}
+
+		// if some tags satisfy the current tag filter condition then we run
+		// the metric through all tag expression tests in order to decide
+		// whether those tags should be part of the final result set
+		if len(metricTags) > 0 {
+			if q.testByAllExpressions(id, def, matchName) {
+				for key := range metricTags {
+					select {
+					case tagCh <- key:
+					case <-stopCh:
+						// if execution of query has stopped because the max tag
+						// count has been reached then tagCh <- might block
+						// because that channel will not be consumed anymore. in
+						// that case the stop channel will have been closed so
+						// we so we exit here
+						break IDS
+					}
+					resultsCache[key] = struct{}{}
+				}
+			} else {
+				// check if we need to stop
+				select {
+				case <-stopCh:
+					break IDS
+				default:
+				}
+			}
+		}
+	}
+
+	q.wg.Done()
+}
+
+// determines whether the given tag prefix/tag match will match the special
+// tag "name". if it does, then we can omit some filtering because we know
+// that every metric has a name
+func (q *TagQuery) tagFilterMatchesName() bool {
+	matchName := false
+
+	if q.filterTagBy == PREFIX_TAG || q.startWith == PREFIX_TAG {
+		if strings.HasPrefix("name", q.tagPrefix) {
+			matchName = true
+		}
+	} else if q.filterTagBy == MATCH_TAG || q.startWith == MATCH_TAG {
+		if q.tagMatch.value.MatchString("name") {
+			matchName = true
+		}
+	} else {
+		// some tag queries might have no prefix specified yet, in this case
+		// we do not need to filter by the name
+		matchName = true
+	}
+
+	return matchName
+}
+
+// RunGetTags executes the tag query and returns all the tags of the
+// resulting metrics
+func (q *TagQuery) RunGetTags(index TagIndex, byId map[string]*idx.Archive) map[string]struct{} {
+	q.index = index
+	q.byId = byId
+
+	maxTagCount := int32(math.MaxInt32)
```

Comment:
why this and not `var maxTagCount int32` ?

Reply:
ah, probably because you just always check if the current num is smaller than maxTagCount, whether's in been set to the real value yet or not.

Reply:
yes, that ^^